File: pc/src/renderer/prepareVpnConnect.ts
import api from './api'
import { applyWorkerCount } from './hashChannelHelper'
import {
  activeServerHashes,
  mapHashesResponse,
  saveHashItems,
  getSavedHashItems,
  type HashItem,
} from './hashItemsStore'
import type { VpnConfigPayload } from './vkConfig'
import { pushLog } from './debugLog'

/** Перед vpnConnect: слоты на сервере, свежие хеши, максимум потоков для libclient. */
export async function prepareVpnConnectConfig(
  config: VpnConfigPayload,
  fingerprint: string,
): Promise<VpnConfigPayload> {
  let merged = { ...config }

  try {
    await api.post('/api/vpn/connect', {
      device_fingerprint: fingerprint,
      device_type: 'pc',
    })
  } catch {
    /* connect может вернуть 403 при лимите устройств — хеши всё равно пробуем */
  }

  let items: HashItem[] = getSavedHashItems()
  try {
    const hashesRes = await api.get('/api/vpn/hashes')
    const downloaded = mapHashesResponse(hashesRes.data)
    if (downloaded.length > 0) {
      saveHashItems(downloaded)
      items = downloaded
    }
    const active = activeServerHashes(items).length
    if (active < 4) {
      try {
        await api.post('/api/vpn/hashes/request-refresh', { device_fingerprint: fingerprint })
        const again = await api.get('/api/vpn/hashes')
        const refreshed = mapHashesResponse(again.data)
        if (refreshed.length > items.length) {
          saveHashItems(refreshed)
          items = refreshed
        }
      } catch {
        /* AI-агент может быть выключен */
      }
    }
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e)
    pushLog('Main', `hash sync: ${msg}`, 'W')
  }

  const serverHashes = activeServerHashes(items).map(i => i.hash.trim()).filter(Boolean)
  if (serverHashes.length > 0) {
    merged = { ...merged, vk_hashes: serverHashes }
  } else if (merged.vk_hashes?.length) {
    merged = {
      ...merged,
      vk_hashes: merged.vk_hashes.map(h => h.trim()).filter(Boolean),
    }
  }

  try {
    const cfgRes = await api.get(`/api/vpn/config?fingerprint=${encodeURIComponent(fingerprint)}`)
    const fresh = cfgRes.data as VpnConfigPayload
    if (fresh.vk_hashes?.length) {
      merged = {
        ...merged,
        ...fresh,
        vk_hashes: fresh.vk_hashes,
        stream_count: fresh.stream_count,
      }
    }
  } catch {
    /* остаёмся на merged из register + hash sync */
  }

  const prepared = applyWorkerCount(merged)
  pushLog(
    'Main',
    `prepare n=${prepared.stream_count} hashes=${prepared.vk_hashes?.length ?? 0} stream_count_api=${merged.stream_count ?? '?'}`,
  )
  return prepared
}

File: app/src/main/kotlin/com/silent/vpn/data/PrepareVpnConnect.kt
package com.silent.vpn.data

import com.silent.vpn.util.DebugLog

/** Перед VPN: слоты на сервере, свежие хеши, максимум потоков для libclient (как PC). */
suspend fun SilentRepository.prepareVpnConnectConfig(
    config: VpnConfig,
    fingerprint: String,
): VpnConfig {
    var merged = config

    var hashItems = getSavedHashItems()
    runCatching {
        val hres = getApi().getVpnHashes()
        if (hres.isSuccessful) {
            val items = hres.body()?.toHashItems().orEmpty()
            if (items.isNotEmpty()) {
                saveHashItems(items)
                hashItems = items
            }
        }
        val active = hashItems.activeServerHashes().size
        if (active < HashChannelHelper.MAX_HASHES) {
            runCatching { getApi().requestHashRefresh(ConnectRequest(fingerprint, getApiDeviceType())) }
            val again = getApi().getVpnHashes()
            if (again.isSuccessful) {
                val refreshed = again.body()?.toHashItems().orEmpty()
                if (refreshed.size > hashItems.size) {
                    saveHashItems(refreshed)
                    hashItems = refreshed
                }
            }
        }
    }.onFailure {
        DebugLog.w("PrepareVpnConnect", "hash sync: ${it.message}")
    }

    val boot = getBootstrapHash()?.trim().orEmpty()
    val serverFromItems = hashItems.activeServerHashes().map { it.hash.trim() }.filter { it.isNotBlank() }
    if (serverFromItems.isNotEmpty()) {
        merged = merged.copy(vk_hashes = serverFromItems.take(HashChannelHelper.MAX_HASHES))
    } else {
        merged = merged.copy(
            vk_hashes = merged.vk_hashes
                .filter { it.isNotBlank() && it.trim() != boot }
                .distinct()
                .take(HashChannelHelper.MAX_HASHES),
        )
    }

    runCatching {
        val cfg = getApi().getConfig(fingerprint)
        if (cfg.isSuccessful) {
            val fresh = cfg.body()!!
            val freshServer = fresh.vk_hashes
                .filter { it.isNotBlank() && it.trim() != boot }
                .distinct()
                .take(HashChannelHelper.MAX_HASHES)
            merged = merged.copy(
                device_id = fresh.device_id,
                wg_private_key = fresh.wg_private_key,
                wg_address = fresh.wg_address,
                wg_dns = fresh.wg_dns,
                server_ip = fresh.server_ip,
                server_port = fresh.server_port,
                server_public_key = fresh.server_public_key,
                wdtt_password = fresh.wdtt_password,
                vk_hashes = freshServer.ifEmpty { merged.vk_hashes },
                stream_count = fresh.stream_count,
            )
        }
    }.onFailure {
        DebugLog.w("PrepareVpnConnect", "config refresh: ${it.message}")
    }

    val activeCount = maxOf(
        merged.vk_hashes.size,
        hashItems.activeServerHashes().size,
        1,
    ).coerceAtMost(HashChannelHelper.MAX_HASHES)
    val userWorkers = resolveWorkersForLibclient(activeCount)
    val serverWorkers = merged.stream_count.takeIf { it >= HashChannelHelper.WORKERS_PER_GROUP }
    val workers = if (serverWorkers != null && serverWorkers > userWorkers) {
        HashChannelHelper.workersForLibclient(serverWorkers, activeCount)
    } else {
        userWorkers
    }
    merged = merged.copy(stream_count = workers)
    DebugLog.i(
        "PrepareVpnConnect",
        "n=$workers hashes=${merged.vk_hashes.size} activeHashes=$activeCount",
    )
    return merged
}
